rfc/0003-tests-arch.md

```diff
@@ -0,0 +1,180 @@
+
+1. Setup Kubernetes cluster
+  Kubewarden is expected to run in multiple Kubernetes versions and machine
+  architectures. So, the end-to-end tests should ensure that all the supported
+  versions and architectures are tested.
+2. Install Kuberwarden stack on the cluster from step 1
+  The version to be tested should be installed. This means that when a component
+  is released. It should be tested against all the other components versions which
+  it supposed to work together.
+3. Run basic end-to-end tests
+  Validate if the basic features are working as expected
+4. Run reconfiguration tests
+  Validates if the user is able to reconfigure a Kubewarden stack installed
+5. Run upgrade tests
+  Validate if the user is able to upgrade and continue to work with the latest release.
+
+Steps 3,4 and 5 may run in parallel.
+
+Considering the tests will validate multiple components of the Kubewarden stack
+together, it desirable to keep all the end-to-end test files in a separated
+repository. This repository should contains all the yaml files, scripts, and
+helper code to run the tests. This also helps to avoid duplicating tests that
+validate some feature that impacts multiple components.
+
+The end-to-end tests repository should be independent. Which means that it should
+run the tests independently of the tests and development cycle of the other
+Kubewarden repositories. But every time a new component release is available
+for testing, the end-to-end tests should be trigger automatically with no human
+intervention.  Besides that, running the end-to-end test on demand and in pre defined
+periods is also desirable.
+
+We could also configure the end-to-end tests to run against the latest version of
+all Kubewarden components. Thus, spotting some possible issues before a official
+release candidate.
+
+Considering the high workload nature of the end-to-end tests it may be necessary
+run them in a dedicated runner outside from Github hosted runners. From past
+experiences, the runners from Github could be slow and the tests failed due
+timeout issues.
+
+This tests can be written using bash and frameworks similar to [bats](https://github.com/bats-core/bats-core/).
+
+The developers should be able to run the same end-to-end tests locally with simple
+commands like:
+
+```bash
+# in the end-to-end test respository directory
+make test
+```
+
+## Policy tests
+
+It's also desirable to have tests to validate policies behaviour. So, it is
+necessary provide policy authors some tools to help them testing their policies.
+There is an initial [tool](https://github.com/kubewarden/setup-kubewarden-cluster-action)
+for that. It's a Github action which creates a K3D cluster and installs Kubewarden
+on it.
+
+The policy templates should be updated to have a template workflow file with
+the basic steps to setup the test infrastructure.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* We need considerable initial effort to build the test infrastructure
```

Comment:
One thing that I believe is relevant is that while e2e tests are totally desirable, it's a good approach to keep them to a minimum that is relevant for us.

In general, unit and integration tests, being closer to the code are more explicit about their failures, easier to track what's wrong. e2e tests while checking the "big picture" aren't usually very explicit into what and where went wrong exactly -- and it's fine, is not their goal --.

Just for consideration: https://blog.znsio.com/2011/02/01/inverting-the-testing-pyramid/ in terms of how many and what weight the different test types should ideally have.

Reply:
Yes, I agree. Most of our effort should be in the unit and components tests. That's one of the reasons I consider component tests different from the end-to-end tests. 

However, we should run the basic end-to-end tests to validate the versions work fine together. But it will still require a considerable amount of effort due to the matrix of possibilities that we will validate.